spyderlib/plugins/__init__.py

```diff
@@ -55,6 +57,97 @@ def get_icon(self):
+    """
+    Filter event attached to each QTabBar that holds 2 or more dockwidgets in
+    charge of handling tab rearangement.
+
+    This filter also holds the methods needed for the detection of a drag and
+    the movement of tabs.
+    """
+    def __init__(self, dock_tabbar, main):
+        QObject.__init__(self)
+        self.dock_tabbar = dock_tabbar
+        self.main = main
+        self.moving = False
+        self.start_tab = None
+        self.end_tab = None
+
+    # Helper methods
+    def _get_plugin(self, index):
+        """Get plugin reference based on tab index."""
+        for plugin in self.main.widgetlist:
+            if plugin.get_plugin_title() == self.dock_tabbar.tabText(index):
+                return plugin
+
+    def _get_plugins(self):
+        """
+        Get a list of all the plugins references in the QTabBar to which this
+        event filter was attached.
+        """
+        plugins = []
+        for index in range(self.dock_tabbar.count()):
+            plugin = self._get_plugin(index)
+            plugins.append(plugin)
+        return plugins
+
+    def eventFilter(self,  obj,  event):
+        """Filter mouse press events."""
+        event_type = event.type()
+        if event_type == QEvent.MouseButtonPress:
+            self.tab_pressed(event)
+            return False
+        if event_type == QEvent.MouseMove:
+            self.tab_moved(event)
+            return True
+        if event_type == QEvent.MouseButtonRelease:
+            self.tab_released(event)
+            return True
+        return False
+
+    def tab_pressed(self, event):
+        """Method called when a tab from a QTabBar has been pressed."""
+        if event.button() == Qt.LeftButton:
+            self.start_tab = self.dock_tabbar.tabAt(event.pos())
+
+    def tab_moved(self, event):
+        """Method called when a tab from a QTabBar has been moved."""
+        # If the left button isn't pressed anymore then return
+        if not event.buttons() & Qt.LeftButton:
+            return
+        if not self.moving:
+            QApplication.setOverrideCursor(Qt.ClosedHandCursor)
+            self.moving = True
+
+        pos = event.pos()
+        self.end_tab = self.dock_tabbar.tabAt(pos)
+        if self.end_tab < 0:
+            self.end_tab = self.start_tab
+
+        if self.start_tab != self.end_tab:
+            self.move_tab(self.start_tab, self.end_tab)
+            self.start_tab = self.end_tab
+
+    def tab_released(self, event):
+        """Method called when a tab from a QTabBar has been released."""
+        QApplication.restoreOverrideCursor()
+        self.moving = False
+
+    def move_tab(self, start_tab, end_tab):
+        """Move a tab from a start to and end position."""
+        plugins = self._get_plugins()
+        start_plugin = self._get_plugin(start_tab)
+        end_plugin = self._get_plugin(end_tab)
+
+        start = plugins.index(start_plugin)
+        end = plugins.index(end_plugin)

```

Comment:
What about `start_idx` and `end_idx` for these variable names? I think they should be clearer ;-)